packages/flutter/lib/src/material/time_picker.dart

```diff
@@ -1473,21 +1178,346 @@ class _DialState extends State<_Dial> with SingleTickerProviderStateMixin {
+  const _TimePickerInput({
+    Key key,
+    @required this.initialSelectedTime,
+    @required this.helpText,
+    @required this.onChanged,
+  }) : assert(initialSelectedTime != null),
+        super(key: key);
+
+  /// The time initially selected when the dialog is shown.
+  final TimeOfDay initialSelectedTime;
+
+  /// Optionally provide your own help text to the time picker.
+  final String helpText;
+
+  final ValueChanged<TimeOfDay> onChanged;
+
+  @override
+  _TimePickerInputState createState() => _TimePickerInputState();
+}
+
+class _TimePickerInputState extends State<_TimePickerInput> {
+  TimeOfDay _selectedTime;
+  bool hourHasError = false;
+  bool minuteHasError = false;
+
+  @override
+  void initState() {
+    super.initState();
+    _selectedTime = widget.initialSelectedTime;
+  }
+
+  int _parseHour(String value) {
+    if (value == null || value.isEmpty) {
+      return null;
+    }
+
+    int newHour = int.tryParse(value);
+    if (newHour == null) {
+      return null;
+    }
+
+    if (MediaQuery.of(context).alwaysUse24HourFormat) {
+      if (newHour >= 0 && newHour < 24) {
+        return newHour;
+      }
+    } else {
+      if (newHour > 0 && newHour < 13) {
+        if ((_selectedTime.period == DayPeriod.pm && newHour != 12)
+            || (_selectedTime.period == DayPeriod.am && newHour == 12)) {
+          newHour = (newHour + TimeOfDay.hoursPerPeriod) % TimeOfDay.hoursPerDay;
+        }
+        return newHour;
+      }
+    }
+    return null;
+  }
+
+  int _parseMinute(String value) {
+    if (value == null || value.isEmpty) {
+      return null;
+    }
+
+    final int newMinute = int.tryParse(value);
+    if (newMinute == null) {
+      return null;
+    }
+
+    if (newMinute >= 0 && newMinute < 60) {
+      return newMinute;
+    }
+    return null;
+  }
+
+  void _handleHourSavedSubmitted(String value) {
+    final int newHour = _parseHour(value);
+    if (newHour != null) {
+      _selectedTime = TimeOfDay(hour: newHour, minute: _selectedTime.minute);
+      widget.onChanged(_selectedTime);
+    }
+  }
+
+  void _handleHourChanged(String value) {
+    final int newHour = _parseHour(value);
+    if (newHour != null && value.length == 2) {
+      // If a valid hour is typed, move focus to the minute TextField.
+      FocusScope.of(context).nextFocus();
+    }
+  }
+
+  void _handleMinuteSavedSubmitted(String value) {
+    final int newMinute = _parseMinute(value);
+    if (newMinute != null) {
+      _selectedTime = TimeOfDay(hour: _selectedTime.hour, minute: int.parse(value));
+      widget.onChanged(_selectedTime);
+    }
+  }
+
+  void _handleDayPeriodChanged(TimeOfDay value) {
+    _selectedTime = value;
+    widget.onChanged(_selectedTime);
+  }
+
+  String _validateHour(String value) {
+    final int newHour = _parseHour(value);
+    setState(() {
+      hourHasError = newHour == null;
+    });
+    if (newHour == null) {
+      return ''; // Return empty string to indicate an error.
```

Comment:
So the `_validateHour` and `_validateMinute` functions are passed into the `TextFormField` validator param and they basically use the presence of non-null strings to indicate an error state. So unfortunately, it has to be done this way.

This error means the user has type something that isn't a valid hour/minute and we want to update the UI to indicate that, so we don't need to assert or throw anything because this is about user-facing errors.